In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
color = sns.color_palette()

%matplotlib inline

### Creating a Dataframe with the file

In [2]:
HB_df = pd.read_csv("hotel_bookings.csv")
HB_df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

In [8]:
HB_df.shape

(119390, 32)

#### The data contains 119390 rows and 32 columns

### Checking the number of nullvalues in the DataFrame

In [3]:
HB_df.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

#### This shows that there are missing values in the following columns: 4 in children, 488 in country, 16340 in agent and 112593 in company

In [4]:
HB_df.isnull().sum().sum()

129425

#### The above code adds up the number of missing values in all columns which results to 129425 missing values 

In [17]:
# Some rows have entreis with 0 adults, 0 children and 0 babies
no_guests = list(HB_df.loc[HB_df["adults"] + HB_df["children"] + HB_df["babies"]==0].index)
no_guests

[2224,
 2409,
 3181,
 3684,
 3708,
 4127,
 9376,
 31765,
 32029,
 32827,
 34849,
 34855,
 37459,
 40392,
 40922,
 41020,
 41056,
 42452,
 42591,
 42871,
 42934,
 47166,
 47180,
 47444,
 49264,
 50701,
 50702,
 54025,
 54037,
 54047,
 54093,
 55809,
 60651,
 61244,
 61245,
 61249,
 62967,
 63796,
 64782,
 65907,
 65908,
 65909,
 65910,
 69668,
 72967,
 72968,
 76620,
 76985,
 77001,
 77003,
 77447,
 77998,
 78257,
 78586,
 78672,
 79077,
 79807,
 79808,
 80072,
 81481,
 82264,
 82265,
 82266,
 82267,
 83506,
 83810,
 84198,
 85490,
 85491,
 85539,
 85931,
 88017,
 88327,
 88328,
 88439,
 89359,
 89360,
 90930,
 91428,
 92814,
 92962,
 93204,
 93320,
 93397,
 93753,
 93754,
 94769,
 94821,
 95544,
 96599,
 96648,
 96810,
 98045,
 98046,
 98492,
 99260,
 99338,
 99774,
 99880,
 100273,
 100487,
 100563,
 100780,
 101329,
 101399,
 101401,
 101794,
 101804,
 101805,
 101957,
 102008,
 102054,
 102055,
 102129,
 102150,
 102155,
 102157,
 102185,
 102258,
 102296,
 102319,
 102611,
 102710,

In [16]:
len(no_guests)

180

#### There are a total of 180 rows with no guests. Now we drop those rows

In [18]:
HB_df.drop(HB_df.index[zero_guests], inplace=True)

In [19]:
HB_df.shape

(119210, 32)

#### Next replace the null values with zero

In [33]:
replace_null = {"children:": 0.0,"country": "Unknown", "agent": 0, "company": 0}
HB_df = HB_df.fillna(replace_null)

# "meal" contains values "Undefined", which is equal to SC.
HB_df["meal"].replace("Undefined", "SC", inplace=True)

In [36]:
# converting the cleaned data to csv
HB_df.to_csv('Hotel_bookings_cleaned_data.csv',index=False)